zkp-component/src/zkp_prover.rs

```diff
@@ -165,45 +134,102 @@ impl<N: Network> ZKProver<N> {
         });
     }
 
-    /// This sends the kill signal to the proof generation process.
-    pub(crate) fn cancel_current_proof_production(&mut self) {
-        self.sender.send(()).unwrap();
+    // Upon every election block, a proof generation process is launched.
+    // The assertion holds true since we should only start generating the next proof after its predecessor's
+    // proof has been pushed into our state.
+    //
+    // Note: The election block stream may have blocks that are too old relative to our zkp state;
+    // thus we will filter those blocks out.
+    fn launch_proof_generation(&mut self, block: MacroBlock) {
+        let zkp_state = self.zkp_state.read();
+        assert!(
+                zkp_state.latest_block_number
+                >= block.block_number() - Policy::blocks_per_epoch(),
+                "The current state (block height: {}) should never lag behind more than one epoch. Current height: {}",
+                zkp_state.latest_block_number,
+                block.block_number(),
+            );
+        if zkp_state.latest_block_number == block.block_number() - Policy::blocks_per_epoch() {
+            let (sender, recv) = channel();
+            self.proof_future = Some(
+                launch_generate_new_proof(
+                    recv,
+                    ProofInput {
+                        block: block.clone(),
+                        latest_pks: zkp_state.latest_pks.clone(),
+                        latest_header_hash: zkp_state.latest_header_hash.clone(),
+                        previous_proof: zkp_state.latest_proof.clone(),
+                        genesis_state: self.genesis_state.clone(),
+                        keys_path: self.keys_path.clone(),
+                    },
+                    self.prover_path.clone(),
+                )
+                .map(|res| res.map(|state| (state, block)))
+                .boxed(),
+            );
+            self.sender = Some(sender);
+        } else {
+            log::debug!(
+                "Won't generate zkp for a block of the past; block height: {}, current height: {}",
+                zkp_state.latest_block_number,
+                block.block_number()
+            );
```

Comment:
I'd use tracing style instead:
```suggestion
            log::debug!(
                block_height = zkp_state.latest_block_number,
                current_height = block.block_number(),
                "Won't generate zkp for a block of the past",
            );
```